src/Compilers/CSharp/Test/Semantic/Semantics/UninitializedNonNullableFieldTests.cs

```diff
@@ -935,29 +935,130 @@ class C<T> where T : struct
+            comp.VerifyDiagnostics(
+                // (5,15): warning CS8618: Non-nullable field 'F1' is uninitialized. Consider declaring the field as nullable.
+                //     private T F1;
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "F1").WithArguments("field", "F1").WithLocation(5, 15),
+                // (6,16): warning CS8618: Non-nullable field 'F2' is uninitialized. Consider declaring the field as nullable.
+                //     private T? F2;
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "F2").WithArguments("field", "F2").WithLocation(6, 16),
+                // (7,15): warning CS8618: Non-nullable property 'P1' is uninitialized. Consider declaring the property as nullable.
+                //     private T P1 { get; }
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "P1").WithArguments("property", "P1").WithLocation(7, 15),
+                // (8,16): warning CS8618: Non-nullable property 'P2' is uninitialized. Consider declaring the property as nullable.
+                //     private T? P2 { get; }
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "P2").WithArguments("property", "P2").WithLocation(8, 16));
+        }
+
+        [Fact]
+        public void GenericType_NotNullConstraint()
+        {
+            var source =
+@"#nullable enable
+#pragma warning disable 0169
+class C<T> where T : notnull
+{
+    private T F1;
+    private T? F2;
+    private T P1 { get; }
+    private T? P2 { get; }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (5,15): warning CS8618: Non-nullable field 'F1' is uninitialized. Consider declaring the field as nullable.
+                //     private T F1;
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "F1").WithArguments("field", "F1").WithLocation(5, 15),
+                // (6,16): warning CS8618: Non-nullable field 'F2' is uninitialized. Consider declaring the field as nullable.
+                //     private T? F2;
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "F2").WithArguments("field", "F2").WithLocation(6, 16),
+                // (7,15): warning CS8618: Non-nullable property 'P1' is uninitialized. Consider declaring the property as nullable.
+                //     private T P1 { get; }
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "P1").WithArguments("property", "P1").WithLocation(7, 15),
+                // (8,16): warning CS8618: Non-nullable property 'P2' is uninitialized. Consider declaring the property as nullable.
+                //     private T? P2 { get; }
+                Diagnostic(ErrorCode.WRN_UninitializedNonNullableField, "P2").WithArguments("property", "P2").WithLocation(8, 16));
+        }
+
+        [Fact]
+        public void GenericType_UnmanagedConstraint()
```

Comment:
❔ Should this be tested with a `default` constraint as well? I believe the result would be most similar to this test, where no warnings are reported for the lack of initialization.

Reply:
How could this be tested with the `default` constraint though? This is about field initialization and given `default` can only occur in a method override there is no way to declare a field of type `T` when `T` has the constraint `default`.